components/storage/blockdevice/COMPONENT_OSPIF/OSPIFBlockDevice.cpp

```diff
@@ -0,0 +1,1908 @@
+{
+    int status = 0;
+    _devices_mutex->lock();
+    if (_number_of_active_ospif_flash_csel >= OSPIF_MAX_ACTIVE_FLASH_DEVICES) {
+        status = -2;
+        goto exit_point;
+    }
+
+    // verify the device is unique(no identical csel already exists)
+    for (int i_ind = 0; i_ind < OSPIF_MAX_ACTIVE_FLASH_DEVICES; i_ind++) {
+        if (_active_ospif_flash_csel_arr[i_ind] == csel) {
+            status = -1;
+            goto exit_point;
+        }
+    }
+
+    // Insert new csel into existing device list
+    for (int i_ind = 0; i_ind < OSPIF_MAX_ACTIVE_FLASH_DEVICES; i_ind++) {
+        if (_active_ospif_flash_csel_arr[i_ind] == NC) {
+            _active_ospif_flash_csel_arr[i_ind] = csel;
+            break;
+        }
+    }
+    _number_of_active_ospif_flash_csel++;
+
+exit_point:
+    _devices_mutex->unlock();
+    return status;
+}
+
+int OSPIFBlockDevice::remove_csel_instance(PinName csel)
+{
+    int status = -1;
+    _devices_mutex->lock();
+    // remove the csel from existing device list
+    for (int i_ind = 0; i_ind < OSPIF_MAX_ACTIVE_FLASH_DEVICES; i_ind++) {
+        if (_active_ospif_flash_csel_arr[i_ind] == csel) {
+            _active_ospif_flash_csel_arr[i_ind] = NC;
+            if (_number_of_active_ospif_flash_csel > 0) {
+                _number_of_active_ospif_flash_csel--;
+            }
+            status = 0;
+            break;
+        }
+    }
+    _devices_mutex->unlock();
+    return status;
+}
+
+/*********************************************************/
+/********** SFDP Parsing and Detection Functions *********/
+/*********************************************************/
+int OSPIFBlockDevice::_sfdp_parse_basic_param_table(Callback<int(bd_addr_t, void *, bd_size_t)> sfdp_reader,
+                                                    sfdp_hdr_info &sfdp_info)
+{
+    uint8_t param_table[SFDP_BASIC_PARAMS_TBL_SIZE]; /* Up To 20 DWORDS = 80 Bytes */
+
+    int status = sfdp_reader(sfdp_info.bptbl.addr, param_table, sfdp_info.bptbl.size);
+    if (status != OSPI_STATUS_OK) {
+        tr_error("Init - Read SFDP First Table Failed");
+        return -1;
+    }
+
+    // Check that density is not greater than 4 gigabits (i.e. that addressing beyond 4 bytes is not required)
+    if (sfdp_detect_addressability(param_table, _sfdp_info.bptbl) < 0) {
+        tr_error("Verify 4byte addressing failed");
+        return -1;
+    }
+
+    if (sfdp_detect_device_density(param_table, _sfdp_info.bptbl) < 0) {
+        tr_error("Detecting device density failed");
+        return -1;
+    }
+
+    // Set Page Size (OSPI write must be done on Page limits)
+    _page_size_bytes = sfdp_detect_page_size(param_table, sfdp_info.bptbl.size);
+
+    if (_sfdp_detect_reset_protocol_and_reset(param_table) != OSPIF_BD_ERROR_OK) {
+        tr_error("Init - Detecting reset protocol/resetting failed");
+        return -1;
+    }
+
+    // Detect and Set Erase Types
+    bool shouldSetQuadEnable = false;
```

Comment:
Octal?

Reply:
No, we keep source code about Quad in OSPI driver, because we hope that OSPI driver can support Quad mode in the future.